src/App.tsx

```diff
@@ -1,26 +1,224 @@
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isDisabledInput, setIsDisabledInput] = useState(false);
+  const [isLoadingTodo, setIsLoadingTodo] = useState<number[]>([]);
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+        setTimeout(() => setErrorMessage(''), 3000);
+      });
+  }, []);
+
+  const preparedTodos = useMemo(() => {
+    if (todoStatus === TodoStatus.active) {
+      return todos.filter(todo => !todo.completed);
+    }
+
+    if (todoStatus === TodoStatus.completed) {
+      return todos.filter(todo => todo.completed);
+    }
+
+    return todos;
+  }, [todos, todoStatus]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  function getActiveTodosArray() {
+    return todos.filter(todo => !todo.completed);
+  }
+
+  function getCompletedTodosArray() {
+    return todos.filter(todo => todo.completed);
+  }
+
+  function addTodo() {
+    const correctTitle = newTodoTitle.trim();
+
+    const newTempTodo = {
+      id: 0,
+      title: correctTitle,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setIsDisabledInput(true);
+    setTempTodo(newTempTodo);
+    setIsLoadingTodo(ids => [...ids, 0]);
+
+    createTodo({ title: correctTitle, userId: USER_ID, completed: false })
+      .then(newTodo => {
+        setTodos(currentTodos => [...currentTodos, newTodo]);
+        setNewTodoTitle('');
+        setIsDisabledInput(false);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+        setTimeout(() => setErrorMessage(''), 3000);
+      })
+      .finally(() => {
+        setIsDisabledInput(false);
+        setTempTodo(null);
+        setIsLoadingTodo(ids => ids.filter(todoId => todoId !== 0));
+      });
+  }
+
+  function deleteTodoId(todoId: number) {
+    setIsLoadingTodo(ids => [...ids, todoId]);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(currentTodos =>
+          currentTodos.filter(todo => todo.id !== todoId),
+        );
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+        setTimeout(() => setErrorMessage(''), 3000);
+      })
+      .finally(() => {
+        setIsLoadingTodo([]);
+      });
+  }
+
+  function updateTodoTitle(todo: Todo) {
+    setIsLoadingTodo(ids => [...ids, todo.id]);
+
+    updateTodo(todo)
+      .then(todoFromServer => {
+        setTodos(currentTodos =>
+          currentTodos.map(currentTodo =>
+            currentTodo.id === todoFromServer.id ? todoFromServer : currentTodo,
+          ),
+        );
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+        setTimeout(() => setErrorMessage(''), 3000);
+      })
+      .finally(() => {
+        setIsLoadingTodo([]);
+      });
+  }
+
+  function updateToggle(toggleTodo: Todo) {
```

Comment:
It is important to follow one way of creating functions. fix it everywhere
```suggestion
  const updateToggle = (toggleTodo: Todo) => {
```